File: timus/1869/main.py
#!/usr/bin/env python3

import sys
import dataclasses

from typing import Sequence, Final, Tuple

seats_per_carriage: Final = 36


@dataclasses.dataclass
class Interval:
    first: int
    last: int
    tickets: int


@dataclasses.dataclass
class Route:
    intervals: Sequence[Interval]

    @classmethod
    def build(cls, stations: int):
        intervals = [
            Interval(first=i, last=i + 1, tickets=0) for i in range(1, stations)
        ]
        return cls(intervals=intervals)

    def register(self, first: int, last: int, tickets: int):
        start = min(first, last)
        end = max(first, last)
        for i in range(start, end):
            # print(i, first, last, tickets)
            self.intervals[i].tickets += tickets
        # print(self.intervals)

    @property
    def max_tickets_on_interval(self) -> int:
        return max((i.tickets for i in self.intervals))


@dataclasses.dataclass
class TicketMatrix:
    rows: Sequence[Sequence[int]]

    def add_row(self, line: str):
        row = [int(w) for w in line.split()]
        self.rows.append(row)

    def to_routes(self) -> Tuple[Route, Route]:
        size = len(self.rows)
        route_forward = Route.build(size)
        route_backward = Route.build(size)

        for i in range(size):
            for j in range(size):
                tickets = self.rows[i][j]
                if i < j:
                    # print(">> FORWARD", i, j, tickets)
                    route_forward.register(i, j, tickets)
                elif i > j:
                    # print(">> BACKWARD", i, j, tickets)
                    route_backward.register(i, j, tickets)

        return (route_forward, route_backward)


def main():
    _ = int(sys.stdin.readline())

    ticket_matrix = TicketMatrix([])

    for line in sys.stdin.readlines():
        ticket_matrix.add_row(line)

    (route_forward, route_backward) = ticket_matrix.to_routes()

    # print(route_forward)
    # print(route_backward)

    max_tickets = max(
        route_forward.max_tickets_on_interval,
        route_backward.max_tickets_on_interval,
    )

    if max_tickets % seats_per_carriage == 0:
        print(int(max_tickets / seats_per_carriage))
    else:
        print(int(max_tickets / seats_per_carriage) + 1)


if __name__ == "__main__":
    main()

File: timus/1869/main.cpp
#include <cassert>
#include <cstddef>
#include <iostream>
#include <sstream>
#include <algorithm>

#include <utility>
#include <vector>

constexpr std::size_t seatsPerCarriage = 36;

struct TInterval {
    std::size_t first;
    std::size_t last;
    std::size_t tickets;
};

class TRoute {
public:
    TRoute(std::size_t stations) {
        intervals.reserve(stations);
        for (std::size_t i = 1; i <= stations; i++) {
            intervals.emplace_back(TInterval{i, i + 1, 0});
        }
    }

    void Register(std::size_t first, std::size_t last, std::size_t tickets) {
        std::size_t start = std::min(first, last);
        std::size_t end = std::max(first, last);
        for (std::size_t i = start; i < end; i++) {
            intervals[i].tickets += tickets;
        }
    }

    std::size_t MaxTicketsPerInterval() const {
        auto iter = std::max_element(intervals.cbegin(), intervals.cend(),
                                     [](const TInterval& a, const TInterval& b) { return a.tickets < b.tickets; });
        return iter->tickets;
    }

private:
    std::vector<TInterval> intervals;
};

class TTicketMatrix {
public:
    TTicketMatrix(std::size_t stations) :Stations(stations) {
        Rows.reserve(stations);
    }

    void AddRow(const std::string& line) {
        std::string target;
        std::stringstream ss(line);
        std::vector<std::size_t> row;
        row.reserve(Stations);

        while (std::getline(ss, target, ' ')) {
            std::size_t value = std::stoi(target);
            row.push_back(value);
        }

        Rows.emplace_back(std::move(row));
    }

    std::pair<TRoute, TRoute> ToRoutes() const {
        TRoute routeForward(Stations);
        TRoute routeBackward(Stations);

        assert(Rows.size() == Stations);
        for (std::size_t i = 0; i < Stations; i++) {
            assert(Rows[i].size() == Stations);
            for (std::size_t j = 0; j < Stations; j++) {
                auto tickets = Rows[i][j];
                if (i < j) {
                    routeForward.Register(i, j, tickets);
                } else if (i > j) {
                    routeBackward.Register(i, j, tickets);
                }
            }
        }

        return std::make_pair(std::move(routeForward), std::move(routeBackward));
    }

private:
    std::size_t Stations;
    std::vector<std::vector<std::size_t>> Rows;
};

int main() {
    std::size_t stations;
    std::cin >> stations;

    TTicketMatrix ticketMatrix(stations);

    std::string target;
    while (std::getline(std::cin, target)) {
        if (target.size()) {
            ticketMatrix.AddRow(target);
        }
    }

    auto result = ticketMatrix.ToRoutes();
    auto maxTickets = std::max(result.first.MaxTicketsPerInterval(), result.second.MaxTicketsPerInterval());

    if (maxTickets % seatsPerCarriage == 0) {
        std::cout << maxTickets / seatsPerCarriage << std::endl;
    } else {
        std::cout << maxTickets / seatsPerCarriage + 1 << std::endl;
    }
}
